### PR TITLE
Einsum Layer Performance Test

### PR DESCRIPTION
**Pull request by Abdurrahheem**
## This PR adds performance tests for Einsum Layer. See below results of performance test on different inputs

**Notation:**
- WX: windows10_x64
- MX: macos_x64
- MA: macos_arm64
- UX: ubuntu_x64
- UA: ubuntu_arm64

All data in ms (milliseconds).
Gemm is backend for matrix multiplication

---

Benchmarks:

| Equation                | Inputs Mat Dims                   | UX (ms)        | UA (ms) | MX (ms) | MA (ms) | WX (ms) |
|-------------------------|-----------------------------------|----------------|---------|---------|---------|---------|
| "ij, jk -> ik"          | [2, 3], [3,2]                     | 0.04 ± 0.00    | -       | -       | -       | -       |
| "ij, jk -> ik"          | [20, 30], [30,20]                 | 0.08 ± 0.00    | -       | -       | -       | -       |
| "ij, jk -> ik"          | [113, 127], [127,113]             | 2.41 ± 0.05    | -       | -       | -       | -       |
| "imkj, injs -> imnks"   | [1, 4, 7, 9], [1, 5, 9, 8]        | 0.11 ± 0.00    | -       | -       | -       | -       |
| "imkj, injs -> imnks"   | [1, 4, 70, 90], [1, 5, 90, 80]    | 15.49 ± 0.46   | -       | -       | -       | -       |
| "imkj, injs -> imnks"   | [1, 4, 73, 91], [1, 5, 91, 57]    | 11.53 ± 0.06   | -       | -       | -       | -       |
| "ij -> i"               | [30, 40]                          | 0.03 ± 0.00    | -       | -       | -       | -       |
| "ij -> i"               | [113, 374]                        | 0.13 ± 0.00    | -       | -       | -       | -       |
| "...ij -> ...i"         | [30, 40]                          | 0.03 ± 0.00    | -       | -       | -       | -       |
| "...ij -> ...i"         | [113, 374]                        | 0.13 ± 0.00    | -       | -       | -       | -       |
| "...ij, ...jk -> ...ik" | [40, 50], [50,80]                 | 0.37 ± 0.01    | -       | -       | -       | -       |
| "...ij, ...jk -> ...ik" | [47, 51], [51, 83]                | 0.43 ± 0.01    | -       | -       | -       | -       |

-----

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [ ] There is a reference to the original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Comment by fengyuentau** [comment]
IIRC, `cv::gemm` is used in einsum layer. Could you use [`fastGemm`](https://github.com/Abdurrahheem/opencv/blob/ash/dev_einsum_pref/modules/dnn/src/layers/cpu_kernels/fast_gemm.hpp) and make a comparison?

**Comment by asmorkalov** [comment]
```
[ RUN      ] Layer_Einsum.einsum/5, where GetParam() = 
Eqiation: 	imkj, injs -> imnks
InputSize: 	2
OutputSize: 	1
InputShape 0: 	100 400 700 900 
InputShape 1: 	100 500 900 800 

[ERROR:0@1279.725] global net_impl.cpp:1197 getLayerShapesRecursively OPENCV/DNN: []:(_input): getMemoryShapes() post validation failed. inputs=2 outputs=2/2 blobs=0 inplace=0
[ERROR:0@1279.728] global net_impl.cpp:1204 getLayerShapesRecursively     input[0] = [ 100 400 700 900 ]
[ERROR:0@1279.728] global net_impl.cpp:1204 getLayerShapesRecursively     input[1] = [ 100 500 900 800 ]
[ERROR:0@1279.728] global net_impl.cpp:1208 getLayerShapesRecursively     output[0] = [ 100 400 700 900 ]
[ERROR:0@1279.728] global net_impl.cpp:1208 getLayerShapesRecursively     output[1] = [ 100 500 900 800 ]
[ERROR:0@1279.728] global net_impl.cpp:1214 getLayerShapesRecursively Exception message: OpenCV(4.8.0-dev) /home/ci/opencv/modules/dnn/src/net_impl.cpp:1193: error: (-2:Unspecified error) in function 'void cv::dnn::dnn4_v20230620::Net::Impl::getLayerShapesRecursively(int, cv::dnn::dnn4_v20230620::Net::Impl::LayersShapesMap&)'
>  (expected: 'total(os[i]) > 0'), where
>     'total(os[i])' is -569803776
> must be greater than
>     '0' is 0

/home/ci/opencv/modules/ts/src/ts_perf.cpp:1965: Failure
Failed
Expected: PerfTestBody() doesn't throw an exception.
  Actual: it throws cv::Exception:
  OpenCV(4.8.0-dev) /home/ci/opencv/modules/dnn/src/net_impl.cpp:1193: error: (-2:Unspecified error) in function 'void cv::dnn::dnn4_v20230620::Net::Impl::getLayerShapesRecursively(int, cv::dnn::dnn4_v20230620::Net::Impl::LayersShapesMap&)'
>  (expected: 'total(os[i]) > 0'), where
>     'total(os[i])' is -569803776
> must be greater than
>     '0' is 0

params    = 
Eqiation: 	imkj, injs -> imnks
InputSize: 	2
OutputSize: 	1
InputShape 0: 	100 400 700 900 
InputShape 1: 	100 500 900 800 

termination reason:  unhandled exception
bytesIn   =          0
bytesOut  =          0
samples   =          0 of 1
outliers  =          0
frequency =          0
[  FAILED  ] Layer_Einsum.einsum/5, where GetParam() = 
Eqiation: 	imkj, injs -> imnks
InputSize: 	2
OutputSize: 	1
InputShape 0: 	100 400 700 900 
InputShape 1: 	100 500 900 800 
 (1288349 ms)
```

**Comment by asmorkalov** [comment]
Abduragim will add fastGemm with the next iteration.

File: modules/dnn/perf/perf_einsum.cpp [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+        outputSize = outputSize_;
+        equation = equation_;
+        einsumInpShapes = einsumInpShapes_;
+    }
+};
+
+static inline void PrintTo(const EinsumParams& params, ::std::ostream* os) {
+     (*os) << "\nEqiation: \t" << params.equation << "\n"
+        << "InputSize: \t" << params.inputSize << "\n"
+        << "OutputSize: \t" << params.outputSize << "\n";
```
Comment (asmorkalov) [inline modules/dnn/perf/perf_einsum.cpp:26]: Please remove \t, \n. `where GetParam() =` goes to performance report and breaks comparison reports. It should be one-line string.

File: modules/dnn/perf/perf_einsum.cpp [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+        for(int i = 0; i < params.einsumInpShapes.size(); i++)
+        {
+            (*os) << "InputShape " << i << ": \t";
```
Comment (asmorkalov) [inline modules/dnn/perf/perf_einsum.cpp:30]: The same here. I propose alternative format like `InputShape={{1, 4,70, 90}, {1, 5, 90, 80}}`

File: modules/dnn/perf/perf_einsum.cpp [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+    {"imkj, injs -> imnks", 2, 1,  {{1, 4, 700, 900}, {1, 5, 900, 800}}},
+
+    {"ij -> i", 1, 1, {{30, 40}}},
+    {"ij -> i", 1, 1, {{300, 400}}},
+
+    {"...ij -> ...i", 1, 1, {{30, 40}}},
+    {"...ij -> ...i", 1, 1, {{300, 400}}},
+
+    {"...ij, ...jk -> ...ik", 2, 1, {{40, 50}, {50, 80}}},
+    {"...ij, ...jk -> ...ik", 2, 1, {{400, 500}, {500, 800}}}
```
Comment (asmorkalov) [inline modules/dnn/perf/perf_einsum.cpp:57]: It makes sense to add couple of odd sizes like 127 or 113. It's usually hardest case for optimization and it shows the worst scenario.

**Review by asmorkalov — APPROVED** [review]
:+1:

**Comment by asmorkalov** [comment]
@dkurt @fengyuentau I want to merge the PR. fastGem will be integrated with the next one to simplify performance comparison. Do you have any concerns?

File: modules/dnn/perf/perf_einsum.cpp [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+    TEST_CYCLE()
+    {
+        net.forward(outputs, "testEinsum");
+    }
+    SANITY_CHECK_NOTHING();
+}
+
+INSTANTIATE_TEST_CASE_P(/**/, Layer_Einsum, testing::ValuesIn(testEinsumConfigs));
+
+}; //namespace
```
Comment (fengyuentau) [inline modules/dnn/perf/perf_einsum.cpp:128]: New line before end of file.

**Review by fengyuentau — COMMENTED** [review]
Also what is the time cost on CI for these tests? is it tolerable (< 1000ms for example)?

**Comment by Abdurrahheem** [comment]
> Also what is the time cost on CI for these tests? is it tolerable (< 1000ms for example)?

@asmorkalov

**Comment by asmorkalov** [comment]
On my old PC without AVX2: 17 tests from 1 test case ran. (15615 ms total)
The longest case is:
```
[ RUN      ] Layer_Einsum.einsum/7, where GetParam() = Eqiation=imkj, injs -> imnks, InputSize=2, OutputSize=1, InputShape={{1, 4, 700, 900}, {1, 5, 900, 800}}
.
[ PERFSTAT ]    (samples=10   mean=1273.04   median=1274.26   min=1261.88   stddev=6.78 (0.5%))
```

**Comment by fengyuentau** [comment]
@Abdurrahheem Could you collect the perf results from [detail pages](https://github.com/opencv/opencv/actions/runs/6653176343/job/18078632250?pr=24445) and fill your table in the first comment?

**Comment by asmorkalov** [comment]
@fengyuentau I propose to rerun the benchmark locally and update the PR. CI runs perf tests with single iteration and concurrently with other builds. The numbers are not reliable.

**Comment by fengyuentau** [comment]
ARM64: ~3.5s

> [ RUN      ] Layer_Einsum.einsum/7, where GetParam() = Eqiation=imkj, injs -> imnks, InputSize=2, OutputSize=1, InputShape={{1, 4, 700, 900}, {1, 5, 900, 800}}
[ PERFSTAT ]    (samples=1   mean=3578.70   median=3578.70   min=3578.70   stddev=0.00 (0.0%))


X64: ~1.8s

> [ RUN      ] Layer_Einsum.einsum/7, where GetParam() = Eqiation=imkj, injs -> imnks, InputSize=2, OutputSize=1, InputShape={{1, 4, 700, 900}, {1, 5, 900, 800}}
[ PERFSTAT ]    (samples=1   mean=1797.72   median=1797.72   min=1797.72   stddev=0.00 (0.0%))

Win-X64: ~7.6s

> [ RUN      ] Layer_Einsum.einsum/7, where GetParam() = Eqiation=imkj, injs -> imnks, InputSize=2, OutputSize=1, InputShape={{1, 4, 700, 900}, {1, 5, 900, 800}}
[ PERFSTAT ]    (samples=1   mean=7660.13   median=7660.13   min=7660.13   stddev=0.00 (0.0%))


I propose to make it a smaller scale.

**Comment by asmorkalov** [comment]
@Abdurrahheem friendly reminder.

**Comment by Abdurrahheem** [comment]
> @fengyuentau I propose to rerun the benchmark locally and update the PR. CI runs perf tests with single iteration and concurrently with other builds. The numbers are not reliable.

I am only able to test on Ubuntu locally current due to lack of different platforms

**Comment by Abdurrahheem** [comment]
Updated the table with performance results.

**Review by fengyuentau — APPROVED** [review]
Thank you!

**Pull request merged**